database/pins-repository.go

```diff
@@ -0,0 +1,265 @@
+package database
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/moby/locker"
+	"github.com/threefoldtech/tf-pinning-service/pinning-api/models"
+	"gorm.io/gorm"
+)
+
+type pins struct {
+	db    *gorm.DB
+	locks *locker.Locker
+	mu    sync.Mutex
+}
+
+var locks = locker.New()
+
+func GetPinsRepository() PinsRepository {
+	return &pins{
+		db:    DB,
+		locks: locks,
+	}
+}
+
+func (r *pins) InsertOrGet(ctx context.Context, user_id uint, pinStatus models.PinStatus) (models.PinStatus, error) {
+	pin := PinDTO{}
+	pin.FromEntity(pinStatus)
+
+	pin.UserID = user_id
+	uuid := pin.UUID
+	pin.UUID = ""
+
+	tx := r.db.Where("user_id = ? AND cid = ?", pin.UserID, pin.Cid).Attrs(PinDTO{UUID: uuid}).FirstOrCreate(&pin)
+	if tx.Error != nil {
+		return models.PinStatus{}, tx.Error
+	}
+
+	return pin.ToEntity(), nil
+}
+
+func (r *pins) Patch(ctx context.Context, user_id uint, id string, fields map[string]interface{}) error {
+	tx := r.db.Model(&PinDTO{}).Where("uuid = ? AND user_id = ?", id, user_id).Updates(fields)
+	if tx.Error != nil {
+		return tx.Error
+	}
+	return nil
+}
+
+func (r *pins) FindByID(ctx context.Context, user_id uint, id string) (models.PinStatus, error) {
+	var pin PinDTO
+	tx := r.db.First(&pin, "uuid = ? AND user_id = ?", id, user_id)
+	if tx.RowsAffected == 0 {
+		return models.PinStatus{}, errors.New("id not exists")
+	}
+	return pin.ToEntity(), nil
+}
+
+func (r *pins) Find(
+	ctx context.Context,
+	user_id uint,
+	cids, statuses []string,
+	name string,
+	before, after time.Time,
+	match string,
+	limit int,
+) (models.PinResults, error) {
+	var pins []PinDTO
+
+	queryDB := r.db.Model(PinDTO{})
+	if len(cids) != 0 {
+		queryDB = queryDB.Where("cid IN ?", cids)
+	}
+	if name != "" {
+		if name != "" {
+			switch match {
+			case string(models.IEXACT):
```

Comment:
Instead of converting string(models.IXACT) would't it be better if the `match` type itself was `TextMatchingStrategy`